core/src/thread_parker/mod.rs

```diff
@@ -0,0 +1,86 @@
+use cfg_if::cfg_if;
+use std::time::Instant;
+
+/// Trait for the platform thread parker implementation.
+///
+/// All unsafe methods are unsafe because the Unix thread parker is based on
+/// pthread mutexes and condvars. Those primitives must not be moved and used
+/// from any other memory address than the one they were located at when they
+/// were initialized. As such, it's UB to call any unsafe method on
+/// `ThreadParkerT` if the implementing instance has moved since the last
+/// call to any of the unsafe methods.
```

Comment:
It's kind of sad how this one implementation spreads `unsafe` to all platforms and the API in general. If the generic Unix implementation just boxed the mutex/condvar I think it could become safe, right? It might lose some performance, but that can be measured so we have some numbers. If the difference is not too large, maybe it would be sane to make the entire thread parker API safe in this way?

We can probably move more platforms away from the generic Unix parker and to better, more specific parkers. For example, could not Android use the same futex based one as Linux? Is there nothing more modern than pthread mutexes for BSD?

Reply:
I believe Android should already be using futexes? It uses the Linux kernel so `target_os` should be `linux`, right?

Reply:
No, it has `target_os = "android"`. See here: https://github.com/rust-lang/libc/blob/master/src/unix/mod.rs#L1141

Reply:
Maybe use [`Pin`](https://doc.rust-lang.org/stable/std/pin/) to guarantee that self is not moved?

Reply:
@bjorn3 Do you have an example on how to do that? I don't know a way to express that without introducing a layer of indirection between the code holding the `ThreadParker` and the actual pthread C structs.

Reply:
Note that there's a lot of UB scenarios with pthread utilities that are beyond movement (that's one of the major reasons for moving to parking_lot!). I suspect that this is is probably going to have to stay `unsafe` even if we were to box primitives, but it would be good to document all the unsafe guarantees we know of at least that must be upheld.

While it's unfortuante that the `unsafe` leaks to other platforms it's also sort of the reality of writing portable code. You're often limited by the lowest common denominator.